crates/telio-traversal/src/endpoint_state.rs

```diff
@@ -0,0 +1,120 @@
+use std::fmt;
+use telio_utils::telio_log_warn;
+
+#[derive(Debug, Default, PartialEq)]
+pub enum EndpointState {
+    #[default]
+    Disconnected,
+    EndpointGathering,
+    Ping,
+    Published,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Event {
+    SendCallMeMaybeRequest,
+    ReceiveCallMeMaybeResponse,
+    Publish,
+    Timeout,
+    EndpointGone,
+}
+
+#[derive(Default, Debug)]
+pub struct EndpointStateMachine {
+    state: EndpointState,
+    last_event: Option<Event>,
+}
+
+impl EndpointStateMachine {
+    pub fn handle_event(&mut self, event: Event) {
+        match (&self.state, &event) {
+            (EndpointState::Disconnected, Event::SendCallMeMaybeRequest) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::EndpointGathering;
+            }
+
+            (EndpointState::EndpointGathering, Event::ReceiveCallMeMaybeResponse) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Ping;
+            }
+
+            (EndpointState::Ping, Event::Publish) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Published;
+            }
+
+            (EndpointState::EndpointGathering, Event::Timeout)
+            | (EndpointState::Ping, Event::Timeout) => {
```

Comment:
I hope I'm getting the syntax right here, but you're supposed to be able to do something like this
```suggestion
            (EndpointState::EndpointGathering|EndpointState::Ping, Event::Timeout) => {
```

Reply:
you're right https://play.rust-lang.org/?version=stable&mode=debug&edition=2021

:D